src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -144,17 +147,45 @@ protected override bool TryParseRequest(ReadResult result, out bool endConnectio
             }
         }
 
+        public bool TryUpdateOutputWindow(int bytes)
+        {
+            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+        }
+
         public override void Abort(ConnectionAbortedException abortReason)
         {
+            if (Interlocked.Exchange(ref _requestAborted, 1) != 0)
+            {
+                return;
+            }
+
             base.Abort(abortReason);
 
             // Unblock the request body.
             RequestBodyPipe.Writer.Complete(new IOException(CoreStrings.Http2StreamAborted, abortReason));
         }
 
-        public bool TryUpdateOutputWindow(int bytes)
+        protected override void ApplicationAbort()
         {
-            return _context.FrameWriter.TryUpdateStreamWindow(_outputFlowControl, bytes);
+            Log.ApplicationAbortedConnection(ConnectionId, TraceIdentifier);
```

Comment:
Should this be logged if the stream was already aborted? Why isn't this combined with ResetAndAbort()?

Reply:
I separated ResetAndAbort because I expect it to be called for other conditions like my next PR for invalid request data. There will be different log statements.

Reply:
I'd combine the methods until there's another caller, but no big deal.

I'm not sure I get the point about there being different log statements. I'm just pointing out it seems wrong to log that app aborted the stream if the stream was already aborted.

Reply:
True, ApplicationAbortedConnection could be logged multiple times, but only for calls to HttpContext.Abort, not for any other reason. That's actually the same as it used to be if you look at the base class (HTTP/1.1) behavior we're overriding. Redundant but harmless.

Reply:
OK